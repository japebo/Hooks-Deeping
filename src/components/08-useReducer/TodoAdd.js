import React from 'react'
import { useForm } from '../../hooks/useForm';


export const TodoAdd = ({handleAdd}) => {
    const [ formValue, handleInputChange, reset ] = useForm({
        task: ''
    });

    const handleSubmit = (e) => {
        e.preventDefault();
        
        if ( formValue.task.trim().length <= 1 ) {
            reset();
            return;
        }
        
        handleAdd( {
            id: new Date().getTime(),
            task: formValue.task,
            done: false    
        })

        reset();
    }

    return (
        <>
            <h4> Add Task </h4>

            <form onSubmit={ handleSubmit }>
                <input
                    type="text"
                    name="task"
                    className='form-control'
                    placeholder="Aprender..."
                    value={ formValue.task }
                    autoComplete="off"
                    onChange={ handleInputChange }
                />
                <button
                    type='submit'
                    className='btn btn-outline-primary mt-1 col-12'
                >
                    Add
                </button>
            </form>
        </>
    )
}
