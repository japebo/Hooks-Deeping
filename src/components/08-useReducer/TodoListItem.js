import React from 'react'

export const TodoListItem = ({todo, index, handleDeletion, handleToggle}) => {
    return (
        <li
            key={todo.id}
            className='list-group-item'
        >
            <p 
                className={ `${todo.done===true && 'complete'}` }
                onClick={ () => {handleToggle(todo.id)} } > {index+1}. {todo.task} </p>
            <button 
                className='btn btn-danger'
                onClick={ () => { handleDeletion(todo.id) } }
            >
                Delete
            </button>
        </li>
    )
}
