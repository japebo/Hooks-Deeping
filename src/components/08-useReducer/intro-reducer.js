
const initialState = [{
    id: 1,
    task: 'Botar la basura',
    done: false
}];

const todoReducer = ( state = initialState, action) => {

    if (action?.type === 'add') {
        return [...state, action.payload]
    }

    return state;
}

let todos = todoReducer();

const newTask = {
    id: 2,
    task: 'Buscar el pan',
    done: false
};

const actionToPerform = {
    type: 'add',
    payload: newTask
};

todos = todoReducer( todos, actionToPerform );

console.log(todos);