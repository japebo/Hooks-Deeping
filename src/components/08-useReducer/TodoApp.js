import React, { useEffect, useReducer } from 'react'
import './styles.css'
import { TodoAdd } from './TodoAdd';
import { TodoList } from './TodoList';
import { todoReducer } from './todoReducer';

const init = () => {
    return JSON.parse(localStorage.getItem('todos')) || [];

    // return [{
    //     id: new Date().getTime(),
    //     task: 'Aprender React',
    //     done: false
    // }];
}

export const TodoApp = () => {

    const [todos, dispatch] = useReducer(todoReducer, [], init);

    useEffect(() => {
        localStorage.setItem( 'todos', JSON.stringify(todos));
    }, [todos])


    const handleAdd = (newTodo) => {
        dispatch({
            type: 'add',
            payload: newTodo
        });
    }

    const handleDeletion = ( todoID ) => {
        const action = {
            type: 'delete',
            payload: todoID
        }

        dispatch(action);

    }

    const handleToggle = ( todoID ) => {

        dispatch({
            type: 'toggle',
            payload: todoID
        });

    }

    return (
        <div>
            <h1>ToDoApp ( {todos.length} )</h1>
            <hr/>

            <div className='row'>
                <div className='col-7'>
                    < TodoList todos={todos} handleDeletion={handleDeletion} handleToggle={handleToggle}/>
                </div>

                <div className='col-5'>
                    <TodoAdd handleAdd={handleAdd}/>
                </div>
            </div>
        </div>
    )
}
