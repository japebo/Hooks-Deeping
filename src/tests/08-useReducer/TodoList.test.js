import { shallow } from "enzyme"
import { TodoList } from "../../components/08-useReducer/TodoList";
import { demoTodos } from "../fixtures/demoTodos"

const handleDeletion = jest.fn();
const handleToggle = jest.fn();

const wrapper = shallow(
    <TodoList 
    todos={demoTodos}
    handleDeletion={handleDeletion}
    handleToggle={handleToggle}
    />);

describe('Pruebas en <TodoList />', () => {
    test('Debe mostrarse correctamente', () => {

        expect(wrapper).toMatchSnapshot();
    })
    
    test('Debe tener dos <TodoListItem />', () => {

        expect( wrapper.find('TodoListItem').length ).toBe( demoTodos.length );

        // console.log( wrapper.find('TodoListItem').at(0).props() );

        expect(wrapper.find('TodoListItem').at(0).prop('handleDeletion')).toEqual(expect.any(Function)); //se espera que la propiedad handleDeletion contenga una funci√≥n. 
        
    })
    
})
