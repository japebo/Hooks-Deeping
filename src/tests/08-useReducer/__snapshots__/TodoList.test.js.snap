// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoList /> Debe mostrarse correctamente 1`] = `
<ul
  className="list-group list-group-flush"
>
  <TodoListItem
    handleDeletion={[MockFunction]}
    handleToggle={[MockFunction]}
    index={0}
    key="1"
    todo={
      Object {
        "done": false,
        "id": 1,
        "task": "Aprender React",
      }
    }
  />
  <TodoListItem
    handleDeletion={[MockFunction]}
    handleToggle={[MockFunction]}
    index={1}
    key="2"
    todo={
      Object {
        "done": false,
        "id": 2,
        "task": "Aprender Mongo",
      }
    }
  />
</ul>
`;
